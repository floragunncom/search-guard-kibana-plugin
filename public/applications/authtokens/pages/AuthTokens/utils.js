/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import moment from 'moment';
import { APP_PATH, AUTH_TOKEN_ACTIONS } from '../../utils/constants';

export function toHumanDate(unixDate) {
  return moment(unixDate).format('YYYY-MM-DDTHH:mm');
}

export const getResourceEditUri = (name) =>
  `${APP_PATH.CREATE_AUTH_TOKEN}?id=${encodeURIComponent(name)}&action=${
    AUTH_TOKEN_ACTIONS.READ_TOKEN
  }`;

export function tokensToUiTokens(resources = []) {
  const tokens = [];
  const revokedTokens = [];

  for (const token of resources) {
    if (Number.isFinite(token.revoked_at)) {
      revokedTokens.push(token);
    } else {
      tokens.push(token);
    }
  }

  return { tokens, revokedTokens };
}
