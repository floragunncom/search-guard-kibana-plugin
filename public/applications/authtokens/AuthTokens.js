/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { BehaviorSubject } from 'rxjs';
import { AppNavLinkStatus } from '../../../../../src/core/public';
import { AuthTokensService } from './services';
import { SEARCHGUARD_APP_CATEGORY } from '../../utils/constants';

export class AuthTokens {
  constructor(coreContext) {
    this.coreContext = coreContext;
    this.appUpdater = new BehaviorSubject(() => ({}));
  }

  mount({ core, httpClient, configService }) {
    return async (params) => {
      const [{ renderApp }] = await Promise.all([import('./app'), configService.fetchConfig()]);
      return renderApp({ element: params.element, core, httpClient, configService });
    };
  }

  setupSync({ core, httpClient, configService }) {
    try {
      this.authTokensService = new AuthTokensService(httpClient);

      core.application.register({
        id: 'searchguard-authtokens',
        title: 'Auth Tokens',
        category: SEARCHGUARD_APP_CATEGORY,
        updater$: this.appUpdater,
        mount: this.mount({ core, httpClient, configService }),
      });
    } catch (error) {
      console.error(`AuthTokens: ${error.stack}`);
    }
  }

  start({ configService }) {
    (async () => {
      try {
        if (configService.isLoginPage()) return;

        const isEnabled = await this.authTokensService.hasUserPermissionsToAccessTheApp();

        if (!isEnabled) {
          this.appUpdater.next(() => ({
            navLinkStatus: AppNavLinkStatus.disabled,
            tooltip: 'Auth Tokens disabled',
          }));
        }
      } catch (error) {
        console.error('AuthTokens, start', error);
      }
    })();
  }
}
