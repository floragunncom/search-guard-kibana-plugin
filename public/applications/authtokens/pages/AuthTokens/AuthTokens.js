/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { EuiInMemoryTable, EuiEmptyPrompt, EuiLink, EuiText, EuiSwitch } from '@elastic/eui';
import {
  ContentPanel,
  CancelButton,
  TableDeleteAction,
  CreateButton,
  TableMultiDeleteButton,
} from '../../../components';
import { AuthTokensService } from '../../services';
import {
  authTokensText,
  authTokensDescriptionText,
  noAuthTokensText,
  createdAtText,
  expiresAtText,
  userNameText,
  revokedAtText,
  revokedTokensText,
  createAuthTokenText,
  theFeatureIsDisabledText,
} from '../../utils/i18n/auth_tokens';
import { nameText } from '../../utils/i18n/common';
import { APP_PATH } from '../../utils/constants';
import { useResourcesTable } from '../../hooks';
import { toHumanDate, getResourceEditUri, tokensToUiTokens } from './utils';
import { Context } from '../../Context';

export function TableColumnText({ value, revokedAt, dataTestSubj }) {
  if (!value) return null;
  return (
    <EuiText
      size="s"
      data-test-subj={dataTestSubj}
      color={Number.isFinite(revokedAt) ? 'subdued' : 'default'}
    >
      {value}
    </EuiText>
  );
}

export function TableColumnDate({ value, revokedAt, dataTestSubj }) {
  if (!value) return null;
  return (
    <EuiText
      size="s"
      data-test-subj={dataTestSubj}
      color={Number.isFinite(revokedAt) ? 'subdued' : 'default'}
    >
      {toHumanDate(value)}
    </EuiText>
  );
}

export function TableColumnTokenName({ value, revokedAt, dataTestSubj, onClick }) {
  return (
    <EuiLink
      onClick={onClick}
      data-test-subj={dataTestSubj}
      color={Number.isFinite(revokedAt) ? 'subdued' : 'primary'}
    >
      {value}
    </EuiLink>
  );
}

export function AuthTokens({ history }) {
  const { triggerWarningCallout } = useContext(Context);
  const {
    service,
    isLoading,
    tableError,
    tableResources,
    deleteTableResources,
    tableResourcesSelection,
    setTableResourcesSelection,
  } = useResourcesTable({
    Service: AuthTokensService,
  });

  const [showRevokedTokens, setShowRevokedTokens] = useState(false);
  const [tokens, setTokens] = useState([]);
  const [revokedTokens, setRevokedTokens] = useState([]);

  useEffect(() => {
    init();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tableResources]);

  async function init() {
    const uiTokens = tokensToUiTokens(tableResources);
    setTokens(uiTokens.tokens);
    setRevokedTokens(uiTokens.revokedTokens);
    console.debug('AuthTokens, init, uiTokens', uiTokens);

    try {
      const isEnabled = await service.isServiceEnabled();
      if (!isEnabled) triggerWarningCallout(theFeatureIsDisabledText);
    } catch (error) {
      console.error('AuthTokens, init, isServiceEnabled', error);
    }
  }

  function renderTableToolsRight() {
    return (
      <EuiSwitch
        data-test-subj="sgTableSwitch-RevokedTokens"
        label={revokedTokensText}
        checked={showRevokedTokens}
        onChange={() => {
          setShowRevokedTokens(!showRevokedTokens);
        }}
      />
    );
  }

  function renderTableToolsLeft() {
    if (tableResourcesSelection.length === 0) return null;
    return (
      <TableMultiDeleteButton
        onClick={() => deleteTableResources(tableResourcesSelection.map(({ _id }) => _id))}
        numOfSelections={tableResourcesSelection.length}
      />
    );
  }

  const tableSelection = {
    onSelectionChange: setTableResourcesSelection,
  };

  const tableActions = [
    {
      render: ({ _id }) => {
        return <TableDeleteAction name={_id} onClick={() => deleteTableResources([_id])} />;
      },
    },
  ];

  const tableColumns = [
    {
      field: '_id',
      name: 'Id',
      footer: 'Id',
      align: 'left',
      sortable: false,
      render: (value, { _id, token_name: name, revoked_at: revokedAt }) => (
        <TableColumnText
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-Id-${name}-${_id}`}
        />
      ),
    },
    {
      field: 'token_name',
      name: nameText,
      footer: nameText,
      align: 'left',
      sortable: true,
      render: (value, { _id, revoked_at: revokedAt }) => (
        <TableColumnTokenName
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-TokenName-${_id}`}
          onClick={() => {
            history.push(getResourceEditUri(_id));
          }}
        />
      ),
    },
    {
      field: 'user_name',
      name: userNameText,
      footer: userNameText,
      align: 'left',
      sortable: true,
      render: (value, { _id, revoked_at: revokedAt }) => (
        <TableColumnText
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-UserName-${_id}`}
        />
      ),
    },
    {
      field: 'expires_at',
      name: expiresAtText,
      footer: expiresAtText,
      align: 'left',
      sortable: true,
      render: (value, { _id, revoked_at: revokedAt }) => (
        <TableColumnDate
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-ExpiresAt-${_id}`}
        />
      ),
    },
    {
      field: 'created_at',
      name: createdAtText,
      footer: createdAtText,
      align: 'left',
      sortable: true,
      render: (value, { _id, revoked_at: revokedAt }) => (
        <TableColumnDate
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-CreatedAt-${_id}`}
        />
      ),
    },
  ];

  if (showRevokedTokens) {
    tableColumns.push({
      field: 'revoked_at',
      name: revokedAtText,
      footer: revokedAtText,
      align: 'left',
      sortable: true,
      render: (value, { _id, revoked_at: revokedAt }) => (
        <TableColumnDate
          value={value}
          revokedAt={revokedAt}
          dataTestSubj={`sgTableCol-RevokedAt-${_id}`}
        />
      ),
    });
  } else {
    tableColumns.push({
      width: '5%',
      align: 'right',
      actions: tableActions,
    });
  }

  const tableSearch = {
    box: {
      incremental: true,
    },
    toolsLeft: renderTableToolsLeft(),
    toolsRight: renderTableToolsRight(),
  };

  const tableProps = {
    itemId: '_id',
    items: showRevokedTokens ? revokedTokens : tokens,
    error: tableError,
    loading: isLoading,
    columns: tableColumns,
    search: tableSearch,
    sorting: true,
    pagination: {
      pageIndex: 0,
      totalItemCount: 10,
      pageSize: 50,
    },
    message: <EuiEmptyPrompt title={<h3>{noAuthTokensText}</h3>} titleSize="xs" />,
  };

  if (!showRevokedTokens) {
    tableProps.selection = tableSelection;
    tableProps.isSelectable = true;
  }

  return (
    <ContentPanel
      title={authTokensText}
      description={authTokensDescriptionText}
      actions={[
        <CancelButton onClick={() => history.push(APP_PATH.HOME)} />,
        <CreateButton
          value={createAuthTokenText}
          onClick={() => history.push(APP_PATH.CREATE_AUTH_TOKEN)}
        />,
      ]}
    >
      <EuiInMemoryTable {...tableProps} />
    </ContentPanel>
  );
}

AuthTokens.propTypes = {
  history: PropTypes.object.isRequired,
};
