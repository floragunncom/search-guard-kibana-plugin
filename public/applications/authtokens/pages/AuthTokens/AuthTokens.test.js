/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { TableColumnText, TableColumnTokenName, TableColumnDate } from './AuthTokens';

describe('AuthTokens', () => {
  let renderer;

  beforeEach(() => {
    renderer = new ShallowRenderer();
  });

  test('TableColumnText for revoked token', () => {
    const tokenId = '398uricjejkvh';
    const value = 'value';
    const revokedAt = 1;

    const tree = renderer.render(
      <TableColumnText value={value} revokedAt={revokedAt} dataTestSubj={`tc-${tokenId}`} />
    );
    expect(tree).toMatchSnapshot();
  });

  test('TableColumnText', () => {
    const tokenId = '398uricjejkvh';
    const value = 'value';

    const tree = renderer.render(<TableColumnText value={value} dataTestSubj={`tc-${tokenId}`} />);
    expect(tree).toMatchSnapshot();
  });

  test('TableColumnTokenName for revoked token', () => {
    const tokenId = '398uricjejkvh';
    const tokenName = 'aToken';
    const revokedAt = 1;
    const onClick = jest.fn();

    const tree = renderer.render(
      <TableColumnTokenName
        value={tokenName}
        revokedAt={revokedAt}
        dataTestSubj={`sgTableCol-TokenName-${tokenId}`}
        onClick={onClick}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  test('TableColumnTokenName', () => {
    const tokenId = '398uricjejkvh';
    const tokenName = 'aToken';
    const onClick = jest.fn();

    const tree = renderer.render(
      <TableColumnTokenName
        value={tokenName}
        dataTestSubj={`sgTableCol-TokenName-${tokenId}`}
        onClick={onClick}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  test('TableColumnDate with no value', () => {
    const tokenId = '398uricjejkvh';

    const tree = renderer.render(<TableColumnDate dataTestSubj={`tc-${tokenId}`} />);
    expect(tree).toMatchSnapshot();
  });
});
