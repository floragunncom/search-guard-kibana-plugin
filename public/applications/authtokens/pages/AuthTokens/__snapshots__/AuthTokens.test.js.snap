// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthTokens TableColumnDate with no value 1`] = `null`;

exports[`AuthTokens TableColumnText 1`] = `
<EuiText
  color="default"
  data-test-subj="tc-398uricjejkvh"
  size="s"
>
  value
</EuiText>
`;

exports[`AuthTokens TableColumnText for revoked token 1`] = `
<EuiText
  color="subdued"
  data-test-subj="tc-398uricjejkvh"
  size="s"
>
  value
</EuiText>
`;

exports[`AuthTokens TableColumnTokenName 1`] = `
<EuiLink
  color="primary"
  data-test-subj="sgTableCol-TokenName-398uricjejkvh"
  onClick={[MockFunction]}
>
  aToken
</EuiLink>
`;

exports[`AuthTokens TableColumnTokenName for revoked token 1`] = `
<EuiLink
  color="subdued"
  data-test-subj="sgTableCol-TokenName-398uricjejkvh"
  onClick={[MockFunction]}
>
  aToken
</EuiLink>
`;
